### ISSUE TITLE
Signal handling for graceful shutdown

### ISSUE THREAD

**Issue by torumakabe** [Azure/azure-functions-docker#404]
I use dotnet:3.0-appservice image for golang based custom handler on Linux. I want to handle signals such as SIGINT/SIGTERM on custom handler, but it seems that we cannot do it because of start.sh (PID 1) does not handle these signals.

At the moment, we have to wait 10 seconds and get exit code 137 (SIGKILL) when I stop the container. Not graceful.

So, I manage it with [domb-init](https://github.com/Yelp/dumb-init) as a workaround.

```Dockefile
[snip]
RUN wget -O /usr/local/bin/dumb-init https://github.com/Yelp/dumb-init/releases/download/v1.2.5/dumb-init_1.2.5_x86_64
RUN chmod +x /usr/local/bin/dumb-init

ENTRYPOINT ["/usr/local/bin/dumb-init", "--"]
CMD ["bash", "/azure-functions-host/start.sh"]
```

Could you please consider implementation of handling logic on start.sh or publishing official guidelines? Thanks.

**Comment by torumakabe** [comment]
FYI. The following is an explanation related to the graceful shutdown of AWS ECS. It's beneficial.

https://aws.amazon.com/jp/blogs/containers/graceful-shutdowns-with-ecs/

**Comment by torumakabe** [comment]
I gave up using dumb-init for forwarding signals. Because dumb-init [does not wait](https://github.com/Yelp/dumb-init/blob/master/dumb-init.c#L112) for handling signal on subprocesses. 

Instead, there is a suggestion to modify start.sh to handle signals. The following is a sample implementation.

```
#!/usr/bin/env bash

# trap signals
trap 'trap " " SIGTERM; kill 0; wait' SIGINT SIGTERM

export DOTNET_USE_POLLING_FILE_WATCHER=true

if [ -z $PORT ]; then
  export ASPNETCORE_URLS=http://*:80
else
  export ASPNETCORE_URLS=http://*:$PORT
fi

if [ -z $SSH_PORT ]; then
  export SSH_PORT=2222
fi

if [ "$APPSVC_REMOTE_DEBUGGING" == "TRUE" ]; then
    export languageWorkers__node__arguments="--inspect=0.0.0.0:$APPSVC_TUNNEL_PORT"
    export languageWorkers__python__arguments="-m ptvsd --host localhost --port $APPSVC_TUNNEL_PORT"
fi

sed -i "s/SSH_PORT/$SSH_PORT/g" /etc/ssh/sshd_config

service ssh start

# run background
if [ -f /azure-functions-host/Microsoft.Azure.WebJobs.Script.WebHost ]; then
    /azure-functions-host/Microsoft.Azure.WebJobs.Script.WebHost &
else
    dotnet /azure-functions-host/Microsoft.Azure.WebJobs.Script.WebHost.dll &
fi

# wait
wait
```

I do not want to maintain the script, so could you please consider implementing it like this or a better solution as upstream? Thanks.

**Comment by torumakabe** [comment]
In the end, this implementation worked in the local environment, but not in Azure. Because App Service/Functions [do not send signals before stop/restart](https://github.com/MicrosoftDocs/azure-docs/issues/72606) without in-process model. It kills our container immediately. 

It is a very important issue for maximizing the robustness of our app. Does anyone have the same issue? Is anyone working on it?

**Comment by torumakabe** [comment]
Does anyone watch this issue? As mentioned in the famous [The Twelve-Factor App](https://12factor.net/disposability), signal handling is essential. Many developers refer to The Twelve-Factor App as best practices for Cloud Native App. FYI, AWS published [an excellent doc](https://aws.amazon.com/jp/blogs/containers/developing-twelve-factor-apps-using-amazon-ecs-and-aws-fargate/).

Of course, signal handling is not the only way to improve disposability and resiliency. However, at present, the implementation ways for containerized Azure Functions is unclear.

**Comment by balag0** [comment]
Yes, this is on the roadmap. Assigning this issue to myself for tracking.

**Comment by PehrGit** [comment]
I tried @ToruMakabe's start.sh and it didn't quite work out of the box for me. I wanted specifically to send "kill SIGINT" to the host, not just kill, so as to give the host a chance to gracefully shut down. Just changing the "kill 0" to "kill SIGINT" didn't work for me, as apparently background processes ignore SIGINT (as explained [here](https://stackoverflow.com/a/38673894)). So I came up with the following, maybe it helps:

```shell
#!/usr/bin/env bash

export DOTNET_USE_POLLING_FILE_WATCHER=true

if [ -z $PORT ]; then
  export ASPNETCORE_URLS=http://*:80
else
  export ASPNETCORE_URLS=http://*:$PORT
fi

if [ -z $SSH_PORT ]; then
  export SSH_PORT=2222
fi

if [ "$APPSVC_REMOTE_DEBUGGING" == "TRUE" ]; then
    export languageWorkers__node__arguments="--inspect=0.0.0.0:$APPSVC_TUNNEL_PORT"
    export languageWorkers__python__arguments="-m ptvsd --host localhost --port $APPSVC_TUNNEL_PORT"
fi

sed -i "s/SSH_PORT/$SSH_PORT/g" /etc/ssh/sshd_config

service ssh start

# run background
if [ -f /azure-functions-host/Microsoft.Azure.WebJobs.Script.WebHost ]; then
    { setsid /azure-functions-host/Microsoft.Azure.WebJobs.Script.WebHost; } &
else
    { setsid dotnet /azure-functions-host/Microsoft.Azure.WebJobs.Script.WebHost.dll; }  &
fi

subshell_pid=$!

function trapSigTerm()
{
  pids=$(ps -ax -o ppid,pid --no-headers |
      sed -r 's/^ +//g;s/ +/ /g' |
      grep "^$subshell_pid " | cut -f 2 -d " ");
  for child in $pids;  do
    echo kill -SIGINT "-$child" && kill -SIGINT "-$child"
  done
  wait $subshell_pid
}

trap trapSigTerm SIGINT SIGTERM

# wait
wait
```

**Comment by rhaikonen** [comment]
Also Net6 console app running in App Service Linux container does not get StopSignal. It is handling long-running operations and would be nice to give it i.e. 5 minutes to gracefully shut down, before killing it. Scaling in / out or even slot swap can cause data loss. Is situation better in Container apps or in anywhere in Azure?

**Comment by abtin** [comment]
Hey folks, there is very limited docs on dockerized go functions, hence I post my question here. Please feel free to direct me to the right place. 

I'm trying to figure out why the base image is so bloated? why there are so many dll files in a container that will be running on linux? Is it possible to use a very light base image or even scratch for a function written in go?

**Comment by marnilss** [comment]
This problem is creeping up all over now: 
local development: https://github.com/Azure/azure-functions-core-tools/issues/4284

dotnet worker: https://github.com/Azure/azure-functions-dotnet-worker/issues/2510 

Perhaps some joint effort on this that solves it in a thought-through manner?

The above mentioned init.sh didn't work for us, so we had to modify it to first send a "drain" command to the host and when "drain-mode" was completed, end the host process. A very complicated workaround.

**Comment by marnilss** [comment]
> Yes, this is on the roadmap. Assigning this issue to myself for tracking.

Soon coming up on 3 years for this high-impact issue.

**Comment by DavidVollmers** [comment]
We are at 4 years now I guess...

**Comment by Frulfump** [comment]
> We are at 4 years now I guess...

5 years, 1 month, and 3 days?
